### ISSUE TITLE
How to marginalize rather than condition on variables to make the output of brms marginal_effects literal AME, MER, and MEM

### ISSUE THREAD

**Issue by realkrantz** [paul-buerkner/brms#552]
#### I propose that the function [marginal_effects](https://www.rdocumentation.org/packages/brms/versions/0.9.1/topics/marginal_effects.brmsfit) be able to estimate the following **marginal effects**: ####

 - Marginal effects at representative values (MERs)
 - Marginal effects at means (MEMs)
 - Average marginal effects (AMEs)

#### Quoting  [Leeper 2018](https://cran.r-project.org/web/packages/margins/vignettes/TechnicalDetails.pdf), "MERs calculate the marginal effect of each variable at a particular combination of X values that is theoretically interesting. MEMs calculate the marginal effects of each variable at the means of the covariates. AMEs calculate marginal effects at every observed value of X and average across the resulting effect estimates.

AMEs are particularly useful because — unlike MEMs — produce a single quantity summary that reflects the full distribution of X rather than an arbitrary prediction. Together with MERs, AMEs have the potential to convey a considerable amount of information about the influence of each covariate on the outcome. Because AMEs average across the variability in the fitted outcomes, they can also capture variability better than MEMs. 

[MERs operate] on a modified set of the full dataset wherein particular variables have been replaced by specified values. This is helpful because it allows for calculation of marginal effects for counterfactual datasets (e.g., what if all women were instead men? what if all democracies were instead autocracies? what if all foreign cars were instead domestic?).

While MERs provide a means to understand and communicate model estimates at theoretically important combinations of covariate values, AMEs provide a natural summary measure that respects both the distribution of the original data..." 

#### For the case of logistic regression with a binary predictor, [AMEs ]((http://www3.nd.edu/%7Erwilliam/) ) measure the discrete change, i.e. how do predicted probabilities change as the binary independent variable changes from 0 to 1, averaged over the actually observed cases.

For a continuous predictor, calculus can be used to compute the AME, but [Cameron and Trivedi 2010](https://www3.nd.edu/~rwilliam/stats3/Margins02.pdf) show that they can also be computed manually as follows: 

1. Compute the predicted values using the observed values of the variables. We will call this prediction1.
2. Change one of the continuous independent variables by a very small amount. Cameron and Trivedi suggest using the standard deviation of the variable divided by 1000. We will refer to this as delta.
3. Compute the new predicted values for each case. Call this prediction2.
4. AME can be obtained by xme=(predict2-predict1)/delta
5. Compute the mean value of xme. This is the AME for the variable in question

#### The current issue is related to the issue https://github.com/paul-buerkner/brms/issues/489 and a discussion on [https://discourse.mc-stan.org](https://discourse.mc-stan.org/t/how-to-marginalize-rather-than-condition-on-variables-to-make-the-output-of-brms-marginal-effects-a-literal-ame/6511/2).

Any idea when we will have this feature in brms?

**Comment by paul-buerkner** [comment]
Thanks for opening this issue. If I understand the terms correctly, MEMs is what marginal effects does by default, while MERs can be achieved via the `conditions` argument. It's indeed AMEs which are missing.  

As you will understand, maintaining brms is nothing I get paid for and it is still a second full-time job to me anyway. So I can only tell you it's done when it's done.

**Comment by realkrantz** [comment]
Thanks a lot for this @paul-buerkner. It is understandable. Just a small comment. The MERs in the above packages [[1](https://cran.r-project.org/web/packages/margins/vignettes/Introduction.html),[2](https://cran.r-project.org/web/packages/modmarg/vignettes/usage.html)] predicts and plots differences in probabilities (AME) (one line per binary treatment) across the covariate we are conditioning on, but in `brms` we get two predictions, one for each category of the treatment. Any tip how to get the MERs that reflect differences in probabilities (AMEs) by changing the binary treatment from 0 to 1, across the covariate we are conditioning on?

**Comment by paul-buerkner** [comment]
Do it manually using the `fitted(..., summary = FALSE)` function with new data, which underlies `marginal_effects`. I trust you will figure this out.

**Comment by realkrantz** [comment]
Thanks for this email. Would simply `calculate manually the differences in the probabilities and differences in their respective confidence limits from the data` extracted from the `brms marginal_effects object` be correct? This difference in probabilities is the quantity of interest, but I am not certain about the idea of calculating differences of the confidence limits between the higher and lower levels of the treatment. Any thoughts?

**Comment by mjskay** [comment]
@paul-buerkner Average marginal effects are something I've been contemplating adding more explicit support for in tidybayes, but was holding off on to see if you did something here :). If this is not a high priority for you, I might forge ahead there --- a potential advantage of having this in tidybayes is that the same code should hopefully work on both rstanarm and brms. 

@realkrantz Doing AMEs for categorical variables is fairly straightforward with tidybayes without additional modification, I think. Assuming I understand them correctly, I put together an example [here](http://htmlpreview.github.io/?https://github.com/mjskay/uncertainty-examples/blob/master/marginal-effects_categorical-predictor.html). Doing AMEs for continuous variables looks like it could be more of a pain...

**Comment by paul-buerkner** [comment]
Thanks, @mjskay! It's not a high priority for me not because it's not important, but because I need to figure out many subtleties in the implementation and I currently don't have the time to do it. My concern is that it's not immediately clear to me how to marginalize over more than one (up to all other) predictors without adding a huge computational burden, in particular, if some of them are continues. But I haven't looked into the details too much. So maybe there is an elegant way to do it.

**Comment by realkrantz** [comment]
Thank a lot @mjskay @paul-buerkner. First of all, this is a tremendous work that you are doing. 

I am doing AME and MER for categorical variables. So [this solution](https://htmlpreview.github.io/?https://github.com/mjskay/uncertainty-examples/blob/master/marginal-effects_categorical-predictor.html) is of tremendous value. Further considerations below.

For `MER`, is it possible to calculate directly the quantity `E[y|A=a1, B=b2] - E[y|A=a1, B=b1]` with its confidence limits? This difference in `marginal mean conditional on particular values of our predictors` is the quantity of interest for `MER`. 

For `AME`, is it possible to estimate directly  `E[y|A=a1] - E[y|A=a2]` with its confidence limits? This would give us be the difference between `the average effect for each value of A, marginalizing over B`. This is the quantity of interest for `AME`.

Thank you very much in advance for this.

**Comment by mjskay** [comment]
For both types of mean differences, you should be able to use `compare_levels`. I updated the AME example I sent before to include [an example of compare_levels](https://htmlpreview.github.io/?https://github.com/mjskay/uncertainty-examples/blob/master/marginal-effects_categorical-predictor.html#differences-in-ames).

More info on `compare_levels` is also in the [tidybayes + brms vignette](http://mjskay.github.io/tidybayes/articles/tidy-brms.html#comparing-levels-of-a-factor).

**Comment by realkrantz** [comment]
@mjskay This is great. This allows estimation of `AME` and `MERs` in a smooth manner. Thanks you very much.

**Comment by realkrantz** [comment]
I would appreciate any help on how to set `conditions` when calculating `AME`  using `tidybayes + brms`. Thank you in advance. This is further discussed [here](https://github.com/mjskay/tidybayes/issues/139).

**Comment by paul-buerkner** [comment]
@jackobailey writes in #767:

> For models other than linear regression, interpreting model coefficients is hard. This is often because they refer to abstract scales (logits, probits) rather than that of the outcome we're interested in (probabilities, counts). One way of overcoming this is to compute *average marginal effects*. Reporting these is near standard in some fields (economics, political science).
> 
> @leeper has implemented these for many model types in his [margins](https://github.com/leeper/margins) package. He has also written [a good accompanying paper](https://cran.r-project.org/web/packages/margins/vignettes/TechnicalDetails.pdf) that explains marginal effects and how to calculate them.
> 
> Margins does not work with brms. This is unfortunate, as brms offers unparalleled opportunities when it comes to estimating non-linear models. Thus, I propose that you implement computing average marginal effects in brms.
> 
> I have had to compute marginal effects for brms models for a project that I am working on. [Here's my attempt at a first pass](https://gist.github.com/jackobailey/0982c89326c36b12d6fa6d6f182189be) based on @leeper's code. As you can see, the code is relatively straightforward and, as it relies on the `fitted` function, suitably generic as to be extendable to most model types.

**Comment by statsccpr** [comment]
So I've been following the rabbit hole on 'marginal' effects of regression models related to brms.
The way I see it, there are two paths that are related but nuanced.

I think the past solutions, directly address one path, when you want to average out a categorical quantity

1. start with E{Y | A,B} and output E(Y | A) by 'averaging' over the categorical B using the literal weighted mean.

* These are the examples using the 'margins' package and what is desired in this specific issue
https://cran.r-project.org/web/packages/margins/vignettes/TechnicalDetails.pdf

2. The other path, is when the variable averaged out is not categorical, and you need to average over a distribution.
start with E{Y | A,B} and output E(Y | A) by 'averaging' over the continuous distribution of B  using integration (whether numerical or monte carlo)
* I think this is what @paul-buerkner was thinking of here
https://github.com/paul-buerkner/brms/issues/489#issue-348201143

* A nice worked out example (actually using brms) is here by @rpsychologist
https://rpsychologist.com/GLMM-part1-lognormal

From what I can tell, the margins package address 1 but does not yet address 2
Meaning, if a user has a regression model with 'random effects' and desires output that integrates out the distribution of the random effect, you still need to roll it by hand as in the rpsychologist post, which is a worked out example of paul's simulation draws example

I think more people are going to start demanding this 2nd option
Where people have non brms solutions 
https://github.com/drizopoulos/GLMMadaptive

**Comment by paul-buerkner** [comment]
I agree with your assessment. Actually, we have 3 situations here, as marginizing over continous variables is technically not the same as marginalizing over random effects. The former are "just" data while the later are parameters with corresponding posterior distribution which makes the marginalization computationally much more demanding. 

I think we should adress them in order, that is

1. marginalize over categorical variables (relatively easy)
2. marginalize over continuous variables, perhaps by converting them to categorical variables with sufficient number of categories
3. marginalize over random effects

**Comment by mjskay** [comment]
I think with the posterior package (and especially careful thinking around https://github.com/jgabry/posterior/issues/39) it would be nice to design an interface in posterior that can do these three types of marginalizations for any model that supports posterior's consistent interface for marginal means, posterior predictive distributions, and random effects.

This might require being a bit more prescriptive about arguments to functions like `pp_expect` and/or providing functions that can inspect a model (e.g. get info about predictors, random effects, etc).

**Comment by leeper** [comment]
I don't think I'll ever do (2) in margins as I'm narrowly focused on the use case in (1).

**Comment by paul-buerkner** [comment]
Andew posted a related paper on discourse, which could help a lot: https://discourse.mc-stan.org/t/request-for-average-predictive-comparisons-in-rstanarm-and-brms/12477

**Comment by jackobailey** [comment]
I have a suggestion that might work for factors, but not continuous variables. Still, it would make some good progress.

Rather than try to marginalise out factors post-hoc, could you do a lot of the work under the hood during the fitting process? I'm thinking of something like using cell mean coding or index coding by default when fitting the model, computing the grand mean of each factor which users could condition on, then returning contrast coded output in the summary regardless. Much like how you rescale the intercept under the hood at the moment.

It would be similar to the approach [you discuss here](https://github.com/paul-buerkner/brms/issues/66).

**Comment by paul-buerkner** [comment]
Thanks, I will take a look!

**Comment by paul-buerkner** [comment]
It seems to me that the marginal effects features requested here are actually implemented in the new `brmsmargins` package (https://github.com/JWiley/brmsmargins) developed by @JWiley. Joshua, can you confirm?

**Comment by JWiley** [comment]
@paul-buerkner I believe this should be true for many cases. So long as the correct input data are specified, `brmsmargins` should calculate AMEs, MEMs, and MERs for many models supported by `brms` and I have hopes to add support for more in the future (anyone with specific interests, please [open an issue on brmsmargins](https://github.com/JWiley/brmsmargins/issues). 
I believe uniquely from other options, `brmsmargins` does support integrating over random effects for correct marginal estimates from random effect models (albeit at considerable computational burden).

**Comment by paul-buerkner** [comment]
Perfect! Thank you so much for working on this! Closing this issue now.
